source/logic.rst

```diff
@@ -0,0 +1,1223 @@
+you can find existential quantifiers hidden all over
+if you know how to spot them.
+For example, divisibility is implicitly an "exists" statement.
+
+.. code-block:: lean
+
+    import tactic
+
+    variables {a b c : ℕ}
+
+    -- BEGIN
+    example (divab : a ∣ b) (divbc : b ∣ c) : a ∣ c :=
+    begin
+      cases divab with d beq,
+      cases divbc with e ceq,
+      rw [ceq, beq],
+      use (d * e), ring
+    end
+    -- END
+
+Try proving the following:
+
+.. code-block:: lean
+
+    import tactic
+
+    variables {a b c : ℕ}
+
+    -- BEGIN
+    example (divab : a ∣ b) (divac : a ∣ c) : a ∣ (b + c) :=
+    sorry
+    -- END
+
+For another important example, a function :math:`f : \alpha \to \beta`
+is said to be *surjective* if for every :math:`y` in the
+codomain, :math:`\beta`,
+there is an :math:`x` in the domain, :math:`\alpha`,
+such that :math:`f(x) = y`.
+Notice that this statement includes both a universal
+and an existential quantifier, which explains
+why the next example makes use of both ``intro`` and ``use``.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    open function
+
+    -- BEGIN
+    example {c : ℝ} : surjective (λ x, x + c) :=
+    begin
+      intro x,
+      use x - c,
+      dsimp, ring
+    end
+    -- END
+
+Try this example yourself:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    open function
+
+    -- BEGIN
+    example {c : ℝ} (h : c ≠ 0) : surjective (λ x, c * x) :=
+    sorry
+    -- END
+
+You can use the theorem ``div_mul_cancel``.
+The next example uses a surjectivity hypothesis
+by applying it to a suitable value.
+Note that you can use ``cases`` with any expression,
+not just a hypothesis.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    open function
+
+    -- BEGIN
+    example {f : ℝ → ℝ} (h : surjective f) : ∃ x, (f x)^2 = 4 :=
+    begin
+      cases h 2 with x hx,
+      use x,
+      rw hx,
+      norm_num
+    end
+    -- END
+
+See if you can use these methods to show that
+the composition of surjective functions is surjective.
+
+.. code-block:: lean
+
+    import tactic
+
+    open function
+
+    variables {α : Type*} {β : Type*} {γ : Type*}
+    variables {g : β → γ} {f : α → β}
+
+    -- BEGIN
+    example (surjg : surjective g) (surjf : surjective f) :
+      surjective (λ x, g (f x)) :=
+    sorry
+    -- END
+
+
+.. _sequences_and_convergence:
+
+Sequences and Convergence
+-------------------------
+
+We now have enough skills at our disposal to do some real mathematics.
+In Lean, we can represent a sequence :math:`s_0, s_1, s_2, \ldots` of
+real numbers as a function ``s : ℕ → ℝ``.
+Such a sequence is said to *converge* to a number :math:`a` if for every
+:math:`\varepsilon > 0` there is a point beyond which the sequence
+remains within :math:`\varepsilon` of :math:`a`,
+that is, there is a number :math:`N` such that for every
+:math:`n \ge N`, :math:`| s_n - a | < \varepsilon`.
+In Lean, we can render this as follows:
```

Comment:
I would introduce the abbreviations `∀ ε > 0` very quickly, even before displaying `def converges_to`